File: frontend/native-plugins/nfc-presence/android/src/main/java/com/synapse/nfcpresence/NfcPresencePlugin.kt
package com.synapse.nfcpresence

import android.app.Activity
import android.nfc.NfcAdapter
import android.nfc.Tag
import android.nfc.tech.Ndef
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import org.json.JSONObject
import java.util.concurrent.Executors

// NDEF payload format:
// { "v": 1, "bid": "<businessId>", "mid": "<missionId>", "ts": <unixEpoch>, "sig": "<hmac-hex>" }
// Server-side HMAC-SHA256 verification via POST /api/missions/{id}/verify-nfc

@CapacitorPlugin(name = "NfcPresence")
class NfcPresencePlugin : Plugin() {

    private var nfcAdapter: NfcAdapter? = null
    private var pendingCall: PluginCall? = null
    private var timeoutRunnable: Runnable? = null
    private val executor = Executors.newSingleThreadExecutor()

    private val readerCallback = NfcAdapter.ReaderCallback { tag ->
        handleTag(tag)
    }

    @PluginMethod
    fun startScan(call: PluginCall) {
        val activity = activity ?: run {
            call.reject("Activity not available.")
            return
        }

        nfcAdapter = NfcAdapter.getDefaultAdapter(context)
        if (nfcAdapter == null || !nfcAdapter!!.isEnabled) {
            call.reject("NFC is not available or disabled on this device.")
            return
        }

        val missionId = call.getString("missionId") ?: run {
            call.reject("missionId is required.")
            return
        }
        val expectedBusinessId = call.getString("expectedBusinessId") ?: run {
            call.reject("expectedBusinessId is required.")
            return
        }
        val timeoutMs = call.getInt("timeoutMs") ?: 30000

        pendingCall = call
        call.setKeepAlive(true)

        val flags = NfcAdapter.FLAG_READER_NFC_A or
                NfcAdapter.FLAG_READER_NFC_B or
                NfcAdapter.FLAG_READER_NFC_F or
                NfcAdapter.FLAG_READER_NFC_V or
                NfcAdapter.FLAG_READER_SKIP_NDEF_CHECK.inv()

        nfcAdapter!!.enableReaderMode(activity, readerCallback, flags, null)

        // Auto-cancel after timeout
        val timeout = Runnable {
            stopReaderMode(activity)
            pendingCall?.reject("Scan timed out.")
            pendingCall = null
        }
        timeoutRunnable = timeout
        android.os.Handler(android.os.Looper.getMainLooper()).postDelayed(timeout, timeoutMs.toLong())
    }

    @PluginMethod
    fun stopScan(call: PluginCall) {
        timeoutRunnable?.let {
            android.os.Handler(android.os.Looper.getMainLooper()).removeCallbacks(it)
        }
        timeoutRunnable = null
        activity?.let { stopReaderMode(it) }
        pendingCall?.reject("Scan cancelled by user.")
        pendingCall = null
        call.resolve()
    }

    private fun handleTag(tag: Tag) {
        timeoutRunnable?.let {
            android.os.Handler(android.os.Looper.getMainLooper()).removeCallbacks(it)
        }
        timeoutRunnable = null
        activity?.let { stopReaderMode(it) }

        executor.execute {
            try {
                val ndef = Ndef.get(tag) ?: run {
                    pendingCall?.reject("Tag does not contain NDEF data.")
                    pendingCall = null
                    return@execute
                }

                ndef.connect()
                val ndefMessage = ndef.cachedNdefMessage ?: ndef.ndefMessage
                ndef.close()

                if (ndefMessage == null || ndefMessage.records.isEmpty()) {
                    pendingCall?.reject("Empty NDEF message on tag.")
                    pendingCall = null
                    return@execute
                }

                val record = ndefMessage.records[0]
                val payload = decodeTextRecord(record.payload) ?: run {
                    pendingCall?.reject("Failed to decode NDEF text record.")
                    pendingCall = null
                    return@execute
                }

                val json = JSONObject(payload)
                val bid = json.optString("bid")
                val mid = json.optString("mid")
                val ts = json.optLong("ts")

                if (bid.isEmpty() || mid.isEmpty()) {
                    pendingCall?.reject("Invalid NDEF payload: missing bid or mid.")
                    pendingCall = null
                    return@execute
                }

                val result = JSObject()
                    .put("verified", true)
                    .put("businessId", bid)
                    .put("missionId", mid)
                    .put("tagTimestamp", ts.toString())
                    .put("rawPayload", payload)

                pendingCall?.resolve(result)
                pendingCall = null

            } catch (e: Exception) {
                pendingCall?.reject("NFC read error: ${e.message}")
                pendingCall = null
            }
        }
    }

    // NDEF Text record: byte[0] = status byte (bit7=UTF16, bits[0..5]=lang length)
    private fun decodeTextRecord(payload: ByteArray): String? {
        if (payload.isEmpty()) return null
        val statusByte = payload[0].toInt() and 0xFF
        val langLength = statusByte and 0x3F
        val isUtf16 = (statusByte and 0x80) != 0
        val textStart = 1 + langLength
        if (textStart >= payload.size) return null
        val textBytes = payload.copyOfRange(textStart, payload.size)
        return if (isUtf16) String(textBytes, Charsets.UTF_16) else String(textBytes, Charsets.UTF_8)
    }

    private fun stopReaderMode(activity: Activity) {
        try { nfcAdapter?.disableReaderMode(activity) } catch (_: Exception) {}
    }
}

File: frontend/native-plugins/nfc-presence/ios/Plugin/NfcPresencePlugin.swift
import Foundation
import Capacitor
import CoreNFC

// NDEF payload format:
// { "v": 1, "bid": "<businessId>", "mid": "<missionId>", "ts": <unixEpoch>, "sig": "<hmac-hex>" }
// HMAC-SHA256(business.NfcSecret, "v=1|bid=<bid>|mid=<mid>|ts=<ts>")
// Verification happens server-side via POST /api/missions/{id}/verify-nfc

@objc(NfcPresencePlugin)
public class NfcPresencePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NfcPresencePlugin"
    public let jsName = "NfcPresence"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "startScan", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopScan", returnType: CAPPluginReturnPromise),
    ]

    private var readerSession: NFCNDEFReaderSession?
    private var pendingCall: CAPPluginCall?
    private var scanOptions: ScanOptions?
    private var timeoutWork: DispatchWorkItem?

    @objc func startScan(_ call: CAPPluginCall) {
        guard NFCNDEFReaderSession.readingAvailable else {
            call.reject("NFC is not available on this device.")
            return
        }

        guard let missionId = call.getString("missionId"),
              let expectedBusinessId = call.getString("expectedBusinessId") else {
            call.reject("missionId and expectedBusinessId are required.")
            return
        }

        let timeoutMs = call.getInt("timeoutMs") ?? 30000

        pendingCall = call
        scanOptions = ScanOptions(missionId: missionId, expectedBusinessId: expectedBusinessId)

        let session = NFCNDEFReaderSession(delegate: self, queue: nil, invalidateAfterFirstRead: true)
        session.alertMessage = "Hold your iPhone near the quest tag at the venue."
        readerSession = session
        session.begin()

        // Auto-cancel after timeout
        let work = DispatchWorkItem { [weak self] in
            self?.readerSession?.invalidate(errorMessage: "Scan timed out.")
            self?.readerSession = nil
            self?.pendingCall?.reject("Scan timed out.")
            self?.pendingCall = nil
        }
        timeoutWork = work
        DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(timeoutMs), execute: work)
    }

    @objc func stopScan(_ call: CAPPluginCall) {
        timeoutWork?.cancel()
        readerSession?.invalidate()
        readerSession = nil
        pendingCall?.reject("Scan cancelled by user.")
        pendingCall = nil
        call.resolve()
    }

    private struct ScanOptions {
        let missionId: String
        let expectedBusinessId: String
    }
}

extension NfcPresencePlugin: NFCNDEFReaderSessionDelegate {
    public func readerSession(_ session: NFCNDEFReaderSession, didInvalidateWithError error: Error) {
        // Session ended — either by user, timeout, or error. pendingCall already resolved/rejected.
    }

    public func readerSession(_ session: NFCNDEFReaderSession, didDetectNDEFs messages: [NFCNDEFMessage]) {
        timeoutWork?.cancel()

        guard let record = messages.first?.records.first,
              record.typeNameFormat == .nfcWellKnown,
              let type = String(data: record.type, encoding: .utf8), type == "T",
              let payload = extractTextPayload(record.payload) else {
            pendingCall?.reject("No readable NDEF Text record found on tag.")
            pendingCall = nil
            return
        }

        guard let data = payload.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let bid = json["bid"] as? String,
              let mid = json["mid"] as? String,
              let ts = json["ts"] as? Int else {
            pendingCall?.reject("Invalid NDEF payload format.")
            pendingCall = nil
            return
        }

        pendingCall?.resolve([
            "verified": true,
            "businessId": bid,
            "missionId": mid,
            "tagTimestamp": String(ts),
            "rawPayload": payload,
        ])
        pendingCall = nil
    }

    // NDEF Text record: language-code prefix (1 byte length + N bytes lang) then text
    private func extractTextPayload(_ data: Data) -> String? {
        guard data.count > 1 else { return nil }
        let langLength = Int(data[0] & 0x3F)
        let textStart = 1 + langLength
        guard textStart < data.count else { return nil }
        return String(data: data[textStart...], encoding: .utf8)
    }
}
